### PR TITLE
benchmarks: Create `benchmarks` directory with example differentiation benchmarks

### PR DESCRIPTION
**Pull request by brocksam**
Adds a new directory, `benchmark`, to contain benchmarks that can be run using [pytest-benchmark](https://pytest-benchmark.readthedocs.io/en/latest/).

Also changed:
- pytest-benchmark is added as a dev dependency.
- pytest is configured to only use `tests` as the default test path. This is to avoid the benchmarks being run when `poetry run pytest` in invoked.
- SymEngine is added as a dev dependency, but with the merging of #39 this can be undone.

I'm opening this as draft so that we can discuss a few things related to how we want to do benchmarking:
- Are we happy with the basic structure of the benchmarks directory, modules, classes, methods etc.?
- Do we want to run these benchmarks in CI? Do we want to do anything more in the short term regarding reporting of the results?

The idea would be to add more benchmark modules with more benchmark classes once we've confirmed the above.

### CODE REVIEW

**Comment by brocksam** [comment]
I struggled with the type hint for the `benchmark` fixture that pytest-benchmark requires to be passed to all benchmarking functions/methods. Mypy is happy with what I've implemented but I think it can definitely be improved as I've used the `...` to avoid properly typing `benchmark` parameter (because I couldn't get it to work otherwise).

**Comment by oscarbenjamin** [comment]
Nice!

I had to fullscreen it to be able to read the output (reads nicer locally with the colouring in the terminal):
```console
$ poetry run pytest benchmarks/
======================================== test session starts ========================================
platform linux -- Python 3.9.16, pytest-7.2.1, pluggy-1.0.0
benchmark: 4.0.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /home/oscar/current/sympy/protosym, configfile: pyproject.toml
plugins: cov-4.0.0, anyio-3.6.2, benchmark-4.0.0, xdoctest-1.1.1, typeguard-2.13.3
collected 6 items                                                                                   

benchmarks/test_differentiation.py ....s.                                                     [100%]


------------------------------------------------------------------- benchmark 'differentiate nested sine first derivative': 3 tests --------------------------------------------------------------------
Name (time in us)            Min                   Max                  Mean              StdDev                Median                IQR            Outliers          OPS            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_symengine            9.2170 (1.0)         58.8760 (1.0)         10.2693 (1.0)        3.3382 (1.0)          9.4020 (1.0)       0.1670 (1.0)         50;96  97,377.5959 (1.0)         466           1
test_simplecas          148.7700 (16.14)      371.6510 (6.31)       154.6296 (15.06)     17.7824 (5.33)       150.9140 (16.05)     2.1828 (13.07)      96;448   6,467.0686 (0.07)       3283           1
test_sympy            1,080.8810 (117.27)   2,039.0740 (34.63)    1,142.2913 (111.23)   148.8303 (44.58)    1,097.5390 (116.73)   41.3020 (247.32)        3;6     875.4334 (0.01)         58           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

------------------------------------------------------------ benchmark 'differentiate nested sine tenth derivative': 2 tests ------------------------------------------------------------
Name (time in ms)          Min                 Max                Mean            StdDev              Median               IQR            Outliers      OPS            Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_simplecas         62.4204 (1.0)       95.9908 (1.0)       65.6069 (1.0)      8.1499 (5.57)      63.2750 (1.0)      1.4775 (1.0)           1;1  15.2423 (1.0)          16           1
test_symengine        324.2816 (5.20)     328.2706 (3.42)     326.4304 (4.98)     1.4624 (1.0)      326.7500 (5.16)     1.7075 (1.16)          2;0   3.0634 (0.20)          5           1
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
=================================== 5 passed, 1 skipped in 6.10s ====================================
```

**Comment by brocksam** [comment]
Interestingly when I run nox locally I get a single xdoctest failure in `protosym/simplecas.py::Expr.eval_f64` where the results differ in the final decimal digit by one. This is not repeated in CI. I can't immediately see anything in the xdoctest documentation about floating point comparison, but I assume it works with [ELLIPSIS](https://docs.python.org/3/library/doctest.html#doctest.ELLIPSIS) like doctest does.

**Comment by oscarbenjamin** [comment]
> Interestingly when I run nox locally I get a single xdoctest failure in `protosym/simplecas.py::Expr.eval_f64` where the results differ in the final decimal digit by one.

I've seen that as well when I ran nox on OSX. This will be a difference in the math module using the underlying C library's `sin` function. I had a look at different options for this.

I am not especially wedded to `xdoctest` (which just came with the cookiecutter template) so we can consider using something else. Currently I don't think that `xdoctest` supports approximate floating point comparisons like SymPy's doctest runner or pytest-doctestplus. We could switch to pytest-doctestplus. There is also numtest (maybe too liberal?):
https://pypi.org/project/numtest/

Another option is writing doctests like:
```python
>>> round(sin(sin(1)), 10)
0.7456241417
```
I imagine that would get tedious though.

**Comment by oscarbenjamin** [comment]
> but I assume it works with [ELLIPSIS](https://docs.python.org/3/library/doctest.html#doctest.ELLIPSIS) like doctest does.

I missed this on a first reading. Maybe that's the quick fix for now. That works in a quick test.

**Comment by oscarbenjamin** [comment]
> * Are we happy with the basic structure of the benchmarks directory, modules, classes, methods etc.?

Yes, I think so.

Ultimately what I would like is a way of writing benchmarks that is somehow agnostic to exactly what system is being used so that the same benchmark specification can be timed against SymPy, SymEngine etc without needing to write separate code for each case. I think that's something that should be outside of protosym though (i.e. in a different repo).

**Comment by oscarbenjamin** [comment]
> Do we want to run these benchmarks in CI?

Yes. A `PR vs main` comparison would be particularly useful. I'm not sure if there's a way to make that "fail" in CI.

> * Do we want to do anything more in the short term regarding reporting of the results?

No, I think that this is fine. Probably once we are ready for a release we can just paste the output into the README or something. For now it will just be useful when exploring different options and we could maybe try to micro-optimise some things.

**Comment by oscarbenjamin** [comment]
> I struggled with the type hint for the `benchmark` fixture that pytest-benchmark requires to be passed to all benchmarking functions/methods

I think it looks fine. Fixture is `Callable[..., T]` where `T` is what the timed function returns. That's all that is needed for mypy to know what the return type is which makes it possible to type check the rest of the function. As of Python 3.10 PEP 612 adds features that can make this more accurate I think:
https://peps.python.org/pep-0612/

With that I think that you can do the more complete:
```python
from typing import ParamSpec, TypeVar

P = ParamSpec('P')
T = TypeVar('T')

Fixture = Callable[[Callable[P, T], P.args, P.kwargs], T]
```
I think that tells the typechecker that the args and kwargs passed to the fixture should be compatible with how you could call the function that is passed as the first argument. It might even be possible then to type the fixture parameter as simple `Fixture` rather than `Fixture[Expr]` because the typechecker could figure out that it is being handed a function that returns `Expr`. Actually it might already be possible to do that with something like:
```
Fixture = Callable[[Callable[..., T], ...], T]
```
I haven't tried using that yet though because we're currently using 3.9 for mypy (mypy with 3.8 already fails).

File: benchmarks/test_differentiation.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+class TestNestedSineFirstDerivative:
+    """Differentiate ``sin(sin(sin(sin(sin(sin(x))))))`` w.r.t. ``x`` once."""
+
+    @staticmethod
+    def test_simplecas(benchmark: Fixture[simplecas.Expr]) -> None:
+        """Differentiate using SimpleCAS."""
+        x = simplecas.Symbol("x")
+        sin = simplecas.sin
+        expr = sin(sin(sin(sin(sin(sin(x))))))
+        benchmark(expr.diff, x)
```
Comment (oscarbenjamin) [inline benchmarks/test_differentiation.py:31]: There should be some check that the output is correct like:
```python
result = benchmark(expr.diff, x)
assert result.eval_f64({x: 1.0}) == ...
```

Comment (oscarbenjamin) [reply benchmarks/test_differentiation.py:31]: The benchmark result should never be shown if the result returned was not checked for correctness in some way.

**Comment by oscarbenjamin** [comment]
In general I think that benchmarking with a cache is tricky. The fairest way would be to run each benchmark in a subprocess with a specified warmup but then only allow a single repeat of the function to be timed. Sort of like this:
```python
import sys
import sympy
from time import time

x = sympy.Symbol('x')

# warmup (needs to be a different expression!):
sympy.exp(x).diff(x)

n = int(sys.argv[1])

# Now the operation
expr = x
for _ in range(n):
    expr = sin(expr)

start = time()
result = expr.diff(x)
end = time()

# assert result.subs(x, 1).evalf() == ...

print(end - start)
```
When I do that with SymPy I get:
```console
$ poetry run python t.py 5
0.009050846099853516
$ poetry run python t.py 5
0.008966684341430664
$ poetry run python t.py 5
0.009221076965332031
```
That's about 9ms. The output from `pytest-benchmark` suggests that it takes 1ms so there is an order of magnitude difference. Doing the same with SymEngine gives about 20us rather than 10us from pytest-benchmark. Of course some of that is the overhead of calling the `time()` function and the Python interpreter executing the three statements, looking up the name `expr` etc. I think it's actually mostly the effect of some sort of cache though.

When you time like above but with protosym you get a time that is the same as the output from pytest-benchmark:
```console
$ poetry run python t.py 5
0.00014138221740722656
$ poetry run python t.py 5
0.00016689300537109375
```
That compares with an average time of 152.9163 from pytest-benchmark. Here the times seem to be equal and as it happens protosym is the only one of these three that does not use a cache.

On the other hand the idea of the cache is that it builds up over time so to some extent cached performance is relevant. We do need to be careful though that we're not just giving meaningless benchmark results (like many SymPy benchmarks).

I don't generally want to introduce a cache in protosym or at least I would like to avoid it as much as possible. The forward graph embodies mcuh of the benefit of a cache but in a better way because:

- It does not overflow during a large operation.
- It does not store anything in memory longer than is needed.

The downside though is that computing the forward graph itself takes time and will probably be a significant portion of the time measured by any benchmark. There might in general be some benefit to caching the forward graph itself e.g. by attaching it as an attribute to a TreeExpr or by having a global weak dict (or even LRU cache). That would make something like this more efficient:
```python
vals = [expr.eval_f64({x: v}) for v in vals]
```
And that is fine if we have a benchmark like that and use it to argue the case. What is not fine though is if our benchmark actually measures unrealistic usage e.g. running the exact same operation many times in a loop is always unrealistic. No one does that in practice or if they do then they should just stop doing it and keep the precomputed result: they should be caching in the calling code rather than requiring all of protosym's core operations to go through a cache.

If there are problems with benchmarking things that are too fast (the reason that repetitions are usually used when benchmarking) then we should just arrange for the test to be something that takes longer e.g. by doing a loop like:
```python
def bench():
    total = 0
    for m in range(n):
        total += sin(sin(sin(sin(sin(x + m)))).diff(x).evalf({x: 1.0})
    return total

result = benchmark(bench)
assert result == ...
```

File: benchmarks/test_differentiation.py (hunk 1 above)
Comment (brocksam) [reply benchmarks/test_differentiation.py]: Valuable point, thanks. Addressed in https://github.com/oscarbenjamin/protosym/pull/40/commits/ddec9f7404520a425bbc23f01d795027b2ad7a13.

**Comment by brocksam** [comment]
> Ultimately what I would like is a way of writing benchmarks that is somehow agnostic to exactly what system is being used so that the same benchmark specification can be timed against SymPy, SymEngine etc without needing to write separate code for each case. I think that's something that should be outside of protosym though (i.e. in a different repo).

I agree. Ideally we'd create something similar to [FPBench](https://fpbench.org) except for CASs. But that's a whole project in amongst itself.

The benchmarks we can create in this repo will be useful for rough demos of ProtoSym/SimpleCAS performance vs SymPy/SymEngine and also for allowing us to A-B test different options for doing the same operation. For example, I could see us  creating a set of benchmark that compares producing LLVM-IR for `lambdify` using a string against using llvmlite's IR API, and the result of that guide whether it's worth making the change.

> In general I think that benchmarking with a cache is tricky.

Your last comment does show that the benchmark results using pytest-benchmark in this PR are somewhat meaningless. Do you think what we have here is of any use or should I pursue setting up a bespoke benchmark runner that does what you've done above with subprocesses to ensure that caching stuff is taken into account fairly?

**Comment by oscarbenjamin** [comment]
> Your last comment does show that the benchmark results using pytest-benchmark in this PR are somewhat meaningless.

I don't think that they are meaningless. At least for one thing this is useful for comparing two different commits of protosym itself or for comparing say different versions of lambdify within protosym. It does mean that care needs to be taken when considering comparisons like sympy vs protosym or symengine.

Also cached time vs non-cached time is tricky because sometimes it is better to compare cached times!

It looks like pedantic mode can be used to make the benchmark only call the function once:
https://pytest-benchmark.readthedocs.io/en/stable/pedantic.html

I don't see any option to run in a subprocess though:
https://github.com/ionelmc/pytest-benchmark/issues/177

Ultimately I want something better that can also compare things that are not Python but for now this is certainly better than nothing. Probably the main thing is to make sure that we are defining the benchmark examples themselves in such a way that we can easily reuse them in a different runner or framework.

File: benchmarks/test_differentiation.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        assert result.eval_f64({x: 1.0}) == pytest.approx(0.13877489681259086)
+
+    @staticmethod
+    def test_sympy(benchmark: Fixture[sympy.Expr]) -> None:
+        """Differentiate using SymPy."""
+        x = sympy.Symbol("x")
+        sin = sympy.sin
+        expr = sin(sin(sin(sin(sin(sin(x))))))
+        result = benchmark(expr.diff, x)
+        assert result.evalf(subs={x: 1.0}) == pytest.approx(0.13877489681259086)
```
Comment (oscarbenjamin) [inline benchmarks/test_differentiation.py:41]: I didn't know about `pytest.approx`

**Comment by oscarbenjamin** [comment]
Unless you're currently working on this I would say that it is good to merge. We can always make further improvements later and for now it's good to have something that we can look at for some timing comparisons even if there are caveats around caching.

**Comment by oscarbenjamin** [comment]
The scheme that I was aiming for in the end sort of like fpbench as you say but I haven't figured out exactly how it would look. It could perhaps be done with yaml files like:
```yaml
--- # calculus1

# Use a setup block to declare what operations are needed from a system that
# could execute this benchmark as well as creating some variables for reuse.

setup:
  variables:
    x: symbol_real
    y: symbol_real
  functions:
    diff: symbolic_diff
    sin: symbolic_sin
    cos: symbolic_cos
    subs: symbolic_subs
    evalf: symbolic_eval_f64

benchmarks:
  - name: trigdiff1
    description: |
      This is an example of simple differentation.

      Here we might have lots of text describing possibly algorithmic details and
      stuff. Note that diff10 makes a very large expression. I wonder if something
      like markdown should be supported here so that it can render nicely like here
      is an equation::

          \frac{d}{dx}\sin^5(x)

    inputs:
      - e: sin(sin(sin(sin(sin(x)))))
        x: x
    time:
      - name: diff
        operation: r = diff(e, x)
        assert_close: evalf(subs(r, x, 1.0)) == 0.1234
      - name: diff2
        operation: r = diff(e, x)
        assert_close: evalf(subs(r, x, 1.0)) == 0.1654
      - name: diff10
        operation: diff(e, x, 10)
        assert_close: evalf(subs(r, x, 1.0)) == 0.5678
```
Then the idea is that this could be parsed and executed or it would be possible to generate code for benchmarks e.g. code to do the above operations with sympy, symengine, Julia, Maxima etc. Then you could use something like this for most test cases and benchmarks. The benchmarks themselves could then also be accessible from outside of Python by implementing an appropriate parser.

There could also then be a way to translate all of the benchmarks into a website where everything gets rendered nicely with latex.

**Comment by oscarbenjamin** [comment]
> I agree. Ideally we'd create something similar to [FPBench](https://fpbench.org) except for CASs. But that's a whole project in amongst itself.

Agreed it would be a lot of work but we could start small here and just use it for some things and then fork it out to a separate project later if it gets good enough.

**Comment by brocksam** [comment]
> Unless you're currently working on this I would say that it is good to merge. We can always make further improvements later and for now it's good to have something that we can look at for some timing comparisons even if there are caveats around caching.

No, I'm not currently working on this so please merge. I will iterate this in future PRs I'm sure and I'll create other PRs for some `evalf` and `lambdify` benchmarks based on your PR benchmarks for those too.

File: benchmarks/test_differentiation.py (hunk 2 above)
Comment (brocksam) [reply benchmarks/test_differentiation.py:41]: I hadn't until recently when I was looking for a non-NumPy way to assert close.

**Comment by oscarbenjamin** [comment]
I'm going to have to write something to make benchmarks for the blog post so maybe that will lead to the start of a different benchmarking framework. It should probably be in a different repo though so that it can be tested against different versions/commits of protosym.

**Comment by oscarbenjamin** [comment]
Looks good. Thanks!

**Pull request merged**